source/store.go

```diff
@@ -16,28 +16,62 @@ limitations under the License.
 
 package source
 
-import "fmt"
+import "errors"
 
-var store = map[string]Source{}
+func init() {
+	// Register all selectable sources under an identifier.
+	RegisterFunc("service", NewServiceSource)
+	RegisterFunc("ingress", NewIngressSource)
+	RegisterFunc("fake", NewFakeSource)
+}
+
+// sourceFunc is a constructor function that returns a Source and an error.
+type sourceFunc func(cfg *Config) (Source, error)
+
+// store is a global store for known sources.
+var store = map[string]sourceFunc{}
+
+// ErrSourceNotFound is returned when a requested source doesn't exist.
+var ErrSourceNotFound = errors.New("source not found")
 
 // Register registers a Source under a given name.
 func Register(name string, source Source) {
+	store[name] = func(_ *Config) (Source, error) { return source, nil }
+}
+
+// RegisterFunc registers a Source under a given name via a constructor function.
+func RegisterFunc(name string, source sourceFunc) {
```

Comment:
I would remove it, and just keep `store["service"] = NewServiceSource` in the `init()`